### PR TITLE
Shared core functionality for messages

### PR DESCRIPTION
**Pull request by dennisafa**
Adds functionality to sleep when no messages are enqueued onto an nf's message ring. 

## Summary:
This is functionality I implemented as part of the larger mTCP project, in which an NF constantly receives lots of messages. This works exactly the same way as sleeping on an empty packet ring. 

**Usage:**
Run the manager with shared CPU enabled.

| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          |  
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                |
| New functionality        | X
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 

## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [x] PR is ready for review

## Test Plan:
Tested by sending a large influx of messages to an NF running in shared cpu mode. Also tested with multiple NF's on the same core. Checked htop CPU usage to verify proper sleeping when no messages were being sent. 

### Review checklist:  

- Sanity checks, assigned to @koolzz  @kevindweb 
    - Run make in `/onvm` and `/examples` directories
    - Test the new functionality

- Code style, assigned to @koolzz  @kevindweb
    - Run linter
    - Check everything style related

- Code design, assigned to @koolzz  @kevindweb
    - Check for memory leaks
    - Check that code is reusable
    - Code is clean, functions are concise
    - Verify that edge cases properly handled

- Performance, assigned to @koolzz  @kevindweb
    - Run Speed Tester NF, report performance.
    - Run pktgen, report performance (not really required)

- Documentation, assigned to @koolzz  @kevindweb
    - Check if the new changes are well documented, in both code and READMEs

### CODE REVIEW

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Error: ERROR: Failed to fetch results from nimbnode30

File: onvm/onvm_nflib/onvm_nflib.c [hunk 1]
```diff
@@ -947,8 +947,13 @@ onvm_nflib_dequeue_messages(struct onvm_nf_local_ctx *nf_local_ctx) {
 
         // Check and see if this NF has any messages from the manager
         if (likely(rte_ring_count(msg_q) == 0)) {
+                if (ONVM_NF_SHARE_CORES) {
+                        rte_atomic16_set(nf->shared_core.sleep_state, 1);
+                        sem_wait(nf->shared_core.nf_mutex);
+                }
                 return;
         }
+
```
Comment (koolzz) [inline onvm/onvm_nflib/onvm_nflib.c]: [nit] rm extra line

**Comment by dennisafa** [comment]
@onvm try it now

**Comment by kevindweb** [comment]
@onvm can we pass this pktgen test?

**Comment by onvm** [comment]
> @onvm can we pass this pktgen test?

## CI Message

Your results will arrive shortly

**Comment by kevindweb** [comment]
Yay to CI! I tested on cloudlab and got 13422053pps for pktgen. 4573120 rx/tx for 2 speed testers together. I'll test the specific message functionality later, but basic performance looks solid

File: onvm/onvm_nflib/onvm_nflib.c [hunk 2]
```diff
@@ -947,6 +947,10 @@ onvm_nflib_dequeue_messages(struct onvm_nf_local_ctx *nf_local_ctx) {
 
         // Check and see if this NF has any messages from the manager
         if (likely(rte_ring_count(msg_q) == 0)) {
+                if (ONVM_NF_SHARE_CORES) {
+                        rte_atomic16_set(nf_local_ctx->nf->shared_core.sleep_state, 1);
+                        sem_wait(nf_local_ctx->nf->shared_core.nf_mutex);
```
Comment (koolzz) [inline onvm/onvm_nflib/onvm_nflib.c]: I'm not sure If I like the fact that it goes to sleep here + when it receives no packets, we should have some sort of benchmarks to verify that. Might be a good quick thing to discuss at the meeting

Comment (dennisafa) [reply onvm/onvm_nflib/onvm_nflib.c]: Yea..maybe we can let the user specify what they want to sleep on (i.e, default for packets)

**Comment by kevindweb** [comment]
@onvm can we show dennis our new features?

**Comment by onvm** [comment]
> @onvm can we show dennis our new features?

## CI Message

Your results will arrive shortly

**Comment by kevindweb** [comment]
@dennisafa pktgen from CI will work if you merge develop into this branch

**Comment by koolzz** [comment]
@onvm perf?

**Comment by kevindweb** [comment]
@onvm perf

**Comment by onvm** [comment]
> @onvm perf

## CI Message

Your results will arrive shortly

**Comment by kevindweb** [comment]
@onvm i believe in you

**Comment by onvm** [comment]
> @onvm i believe in you

## CI Message

Your results will arrive shortly

**Comment by dennisafa** [comment]
@onvm how's it goin pktgen

**Comment by onvm** [comment]
> @onvm how's it goin pktgen

## CI Message

Your results will arrive shortly

**Comment by kevindweb** [comment]
@onvm if you fail I'll catch you now!

**Comment by onvm** [comment]
> @onvm if you fail I'll catch you now!

## CI Message

Your results will arrive shortly

File: onvm/onvm_nflib/onvm_common.h [hunk 3]
```diff
@@ -469,7 +469,7 @@ get_sem_name(unsigned id) {
 
 static inline int
 whether_wakeup_client(struct onvm_nf *nf, struct nf_wakeup_info *nf_wakeup_info) {
-        if (rte_ring_count(nf->rx_q) < PKT_WAKEUP_THRESHOLD)
+        if (rte_ring_count(nf->rx_q) < PKT_WAKEUP_THRESHOLD && rte_ring_count(nf->msg_q) == 0)
```
Comment (koolzz) [inline onvm/onvm_nflib/onvm_common.h]: Might as well make it a macro to stay consistent with the pkt sleep logic, MSG_WAKEUP_THRESHOLD

**Comment by koolzz** [comment]
@dennisafa just add the https://github.com/sdnfv/openNetVM/pull/151/files changes into this pr

**Comment by dennisafa** [comment]
> @dennisafa just add the https://github.com/sdnfv/openNetVM/pull/151/files changes into this pr

oops. got it.

**Comment by onvm** [comment]
> Testing

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> Testing

## CI Message

Error: Failed to parse Speed Tester stats

**Comment by onvm** [comment]
> Testing

## CI Message

Your results will arrive shortly

**Comment by dennisafa** [comment]
awesome, im excited CI has mTCP results now. no real performance benefit,
just the fact that the semwait call is logically equivalent so having it in
both dequeue_messages and dequeue_packets would be extra

**Pull request merged**